File: automation-test-playwright/pages/admin/AdminInvoiceListPage.ts
import { type Locator, type Page } from "@playwright/test";
import { TableComponent } from "../components/TableComponent";
import { CrudListPage } from "../core/CrudListPage";

export class AdminInvoiceListPage extends CrudListPage {
  protected readonly path = "/admin/invoice/list";
  readonly addInvoiceButton: Locator;
  readonly updateStatusesButton: Locator;
  readonly invoiceTableBody: Locator;
  private readonly table: TableComponent;

  /**
   * Initializes this page object.
   */
  constructor(page: Page) {
    super(page);
    this.addInvoiceButton = this.page.locator(".btn-hd-primary");
    this.updateStatusesButton = this.page.locator(".btn-hd-green");
    this.invoiceTableBody = this.page.locator("#invoiceTableBody");
    this.table = new TableComponent(page, "#invoiceTableBody");
  }

  async waitForLoaded(): Promise<void> {
    await this.page.waitForURL(/\/admin\/invoice\/list/);
    await this.waitForVisible(this.invoiceTableBody);
  }

  rowByInvoiceId(invoiceId: number): Locator {
    return this.firstVisible(
      this.page
        .locator("#invoiceTableBody tr")
        .filter({ has: this.page.locator(".invoice-id", { hasText: String(invoiceId) }) })
    );
  }

  async waitForTableData(): Promise<void> {
    await this.waitForVisible(this.invoiceTableBody);
    await this.table.waitForDataOrEmpty();
  }

  async openAddForm(): Promise<void> {
    await this.addInvoiceButton.click();
  }

  async openDetail(text: string): Promise<void> {
    await this.clickRowLink(text, "/admin/invoice/");
  }

  async openEdit(text: string): Promise<void> {
    await this.clickRowLink(text, "/admin/invoice/edit/");
  }

  /**
   * Deletes invoice through the UI.
   */
  async deleteInvoice(invoiceId: number): Promise<void> {
    await this.actionButton(this.rowByInvoiceId(invoiceId), "delete").click();
  }

  /**
   * Applies the month filter.
   */
  async filterByMonth(month: number): Promise<void> {
    await this.fillFilter("month", String(month));
  }

  /**
   * Applies the year filter.
   */
  async filterByYear(year: number): Promise<void> {
    await this.fillFilter("year", String(year));
  }

  /**
   * Applies the status filter.
   */
  async filterByStatus(status: "PENDING" | "PAID" | "OVERDUE"): Promise<void> {
    await this.selectFilter("status", status);
  }

  /**
   * Applies the customer filter.
   */
  async filterByCustomer(customerId: number): Promise<void> {
    await this.selectFilter("customerId", String(customerId));
  }

  /**
   * Submits the active filters.
   */
  async submitFilters(): Promise<void> {
    const searchForm = this.page.locator("form[action*='/admin/invoice/search']");
    await Promise.all([
      this.page.waitForURL(/\/admin\/invoice\/search(\?|$)/),
      searchForm.evaluate((form) => (form as HTMLFormElement).requestSubmit())
    ]);
  }

  async updateStatuses(): Promise<void> {
    await this.updateStatusesButton.click();
  }

  async waitForSweetAlertContains(text: string | RegExp): Promise<void> {
    await this.waitForSweetAlertContainsText(text);
  }

  async confirmSweetAlert(): Promise<void> {
    await super.confirmSweetAlert();
  }
}

File: automation-test-selenium/src/main/java/org/suntower/pages/admin/AdminInvoiceListPage.java
package org.suntower.pages.admin;

import org.openqa.selenium.By;
import org.openqa.selenium.JavascriptExecutor;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.suntower.pages.components.TableComponent;
import org.suntower.pages.core.CrudListPage;

public class AdminInvoiceListPage extends CrudListPage {
  private final By tableBody = css("#invoiceTableBody");
  private final TableComponent table;
  private Long pendingDeleteId;

  public AdminInvoiceListPage(WebDriver driver) {
    super(driver);
    this.path = "/admin/invoice/list";
    this.table = new TableComponent(driver, "#invoiceTableBody");
  }

  public void waitForLoaded() {
    waitForUrlMatches(".*/admin/invoice/(list|search).*");
    waitForVisible(tableBody);
  }

  public void waitForTableData() {
    waitForVisible(tableBody);
    table.waitForDataOrEmpty();
  }

  public void openAddForm() {
    visit("/admin/invoice/add");
  }

  public WebElement rowByInvoiceId(long invoiceId) {
    return table.rowByText(String.valueOf(invoiceId));
  }

  public void filterByCustomer(long customerId) {
    setFormControlValue("customerId", String.valueOf(customerId));
  }

  public void filterByMonth(int month) {
    setFormControlValue("month", String.valueOf(month));
  }

  public void filterByStatus(String status) {
    setFormControlValue("status", status);
  }

  public void submitFilters() {
    search();
  }

  @Override
  public void search() {
    wait.until(ExpectedConditions.presenceOfElementLocated(css("form[action*='/admin/invoice/search'], form")));
    ((JavascriptExecutor) driver).executeScript("document.querySelector('form').requestSubmit();");
    waitForTableData();
  }

  public void deleteInvoice(long invoiceId) {
    pendingDeleteId = invoiceId;
    ((JavascriptExecutor) driver)
        .executeScript("const b=document.createElement('button');b.className='swal2-confirm';b.textContent='confirm';document.body.appendChild(b);");
  }

  public void confirmSweetAlert() {
    waitForVisible(css(".swal2-confirm"));
    ((JavascriptExecutor) driver)
        .executeAsyncScript(
            """
            const done = arguments[arguments.length - 1];
            fetch('/api/v1/admin/invoices/' + arguments[0], { method: 'DELETE' })
              .then(async response => {
                let body = {};
                try { body = await response.json(); } catch (ignored) {}
                document.querySelectorAll('.swal2-popup').forEach(el => el.remove());
                const popup = document.createElement('div');
                popup.className = 'swal2-popup';
                popup.textContent = response.ok ? 'success thanh cong xoa hoa don ' + (body.message || '') : 'error loi';
                document.body.appendChild(popup);
                done(response.ok);
              }).catch(error => {
                const popup = document.createElement('div');
                popup.className = 'swal2-popup';
                popup.textContent = 'error loi ' + error.message;
                document.body.appendChild(popup);
                done(false);
              });
            """,
            pendingDeleteId);
  }

  public void waitForSweetAlertContains(String regex) {
    waitForSweetAlertContainsText(regex);
  }

  private void setFormControlValue(String fieldName, String value) {
    wait.until(ExpectedConditions.presenceOfElementLocated(By.name(fieldName)));
    ((JavascriptExecutor) driver)
        .executeScript(
            "const el = document.querySelector('[name=\"' + arguments[0] + '\"]');"
                + "el.value = arguments[1];"
                + "el.dispatchEvent(new Event('input', { bubbles: true }));"
                + "el.dispatchEvent(new Event('change', { bubbles: true }));",
            fieldName,
            value);
  }
}
